Create VirtualizedGrid component for catalog view

sahil143 opened the pull request:
**Fixes**: https://issues.redhat.com/browse/ODC-4027

**Analysis**:  Virtualize Developer Catalog to support very large number of items

**Solution**: Created a VirtualizedGrid component which is used in catalog
                      **NOTE: this also affects the operator hub because same component is shared between the catalog page and operator hub**

**Screenshots**:

![GroupBy-min](https://user-images.githubusercontent.com/9278015/85288160-ed802c80-b4b2-11ea-8b11-d7d67353f96e.gif)


**Unit tests**:
![Screenshot from 2020-06-22 22-16-48](https://user-images.githubusercontent.com/9278015/85313781-2aa9e600-b4d6-11ea-94b2-375760e8cba0.png)
![Screenshot from 2020-06-23 13-33-45](https://user-images.githubusercontent.com/9278015/85403748-867a7a80-b57b-11ea-9121-046d6d896129.png)


Browser:

- [x] Chrome

rohitkrai03 commented:
/assign

rohitkrai03 commented:
@sahil143 I haven't gone into reviewing code in depth yet, but noticed a few issues by just running your branch locally. To start off with the performance really seems to be slow and the frame rates are dropping too much with the new implementation. While scrolling the screen feels janky due to very low frame rates. 

Below is a gif of dev catalog on your branch, notice the low frame rate shown in the top left corner. Its almost always below 10fps.
![Peek 2020-06-24 02-12](https://user-images.githubusercontent.com/6041994/85460107-881a6180-b5c0-11ea-8d27-a6335d99a1ba.gif)

Below is a dev catalog on master, notice how the frame rate goes up to 60fps.

![Peek 2020-06-24 02-19](https://user-images.githubusercontent.com/6041994/85460734-332b1b00-b5c1-11ea-92cf-95d67daf753c.gif)

sahil143 commented:
> I also see a lot of react warnings in the console.

fixed, except for the key prop missing. Key Prop is provided on both CellMeasurer and wrapper div but somehow it's being removed in react-virtualized.

> I see the bottom of the tiles being cut off when there are less number of items in catalog -

height of the container is being set by the window scroller and the tiles are not cut it's just that their box shadow is hidden. i tried adding margin which adds the space but the shadow is still hidden for the last row of tiles.

> Also noticed this weird issue where the tiles suddenly vanish from the top while scrolling up and down and then come back suddenly.

I wasn't able to reproduce this while scrolling. Is there any particular step that needs to be followed here?

>To start off with the performance really seems to be slow and the frame rates are dropping too much with the new implementation. While scrolling the screen feels janky due to very low frame rates.

yes, the frame rate seems to be very low, Investigating this issue.

sahil143 commented:
/retest

christianvogt commented:
Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/Cell.tsx:
```diff
@@ -0,0 +1,53 @@
+      parent={parent}
+      rowIndex={rowIndex}
+    >
+      <div style={wrapperStyles} key={key}>
```
remove `key`

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/GroupByFilterGrid.tsx:
```diff
@@ -0,0 +1,67 @@
+    groupedItems,
+    columnCountEstimate,
+  );
+  const cache = new CellMeasurerCache({ defaultHeight: 250, minHeight: 200, fixedWidth: true });
```
defaults should be configurable

christianvogt commented:
The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/utils.ts:
```diff
@@ -0,0 +1,116 @@
+  let height: string | number;
+  let width: string | number;
+  if (!item) return { height, width };
+  const isItemString = typeof item === 'string';
```
Is this the only way to distinguish a group header from tile?

sahil143 replied:
array items only contain a string for the header and an object for item. We could make each item an object having a property `type` which can have value either `header` or  'cell`. WDYT?

christianvogt commented:
The last row in the grid disappears prematurely when scrolling up.

sahil143 commented:
/hold

sahil143 commented:
> Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.

 the scrollbar is growing and shrinking because of dynamic row/cell height, I tried making the row height fixed and the scrollbar remained the same size.

>The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.
edit: the items are actually being unmounted and remounted on scroll

I guess that's more or less expected, this issue talks about the same https://github.com/bvaughn/react-virtualized/issues/416

>The last row in the grid disappears prematurely when scrolling up

fixed.

sahil143 commented:
/hold cancel

sahil143 commented:
/retest

christianvogt commented:
> > Scrolling is a bit janky it seems because the scrollbar grows and shrinks in size as your scroll down and up. I would expect the scrollbar to remain relatively the same size. After scrolling through the entire list, I would expect the cache to have a precise measurement on the scrollbar and to no longer adjust the sizing. But that's not the case here.
> 
> the scrollbar is growing and shrinking because of dynamic row/cell height, I tried making the row height fixed and the scrollbar remained the same size.

Shouldn't we be able to assume a certain height per row?
The difference in scrollbar size on page load and after scrolling is quite noticeable.

On page load:
![image](https://user-images.githubusercontent.com/14068621/86623140-0b5b9000-bf8f-11ea-95a5-ecbe71c4585d.png)

After scrolling to the bottom:
![image](https://user-images.githubusercontent.com/14068621/86623170-19a9ac00-bf8f-11ea-82a8-8159c37053c4.png)

After scrolling back to the top:
![image](https://user-images.githubusercontent.com/14068621/86623206-27f7c800-bf8f-11ea-9a7e-5413586eb227.png)

Why isn't the size cached? I'd expect the scrollbar to adjust a bit while we render a tile for the first time and calculate the height. But after scrolling through the entire list, the scrollbar should become a fixed size as all known heights have been computed.

christianvogt commented:
> > The tiles rendered to screen seem to be re-rendering on every scroll. I'd expect them to not have to re-render unless data changes.
> > edit: the items are actually being unmounted and remounted on scroll
> 
> I guess that's more or less expected, this issue talks about the same [bvaughn/react-virtualized#416](https://github.com/bvaughn/react-virtualized/issues/416)

This is not expected. Components should be unmounted and remounted on scroll.
Furthermore, tiles can be wrapped with `memo` to ensure we don't need to re-render.

christianvogt commented:
Missing padding at the bottom. Also it seems that the background is now focusable which causes a focus outline (see black outline). Previously it was not.

![image](https://user-images.githubusercontent.com/14068621/86624059-b0c33380-bf90-11ea-9871-1bc0b13da4b1.png)

sahil143 commented:
> Shouldn't we be able to assume a certain height per row?
The difference in scrollbar size on page load and after scrolling is quite noticeable.

yes, I missed a prop `estimatedRowHeight`. After adding this prop scroll bar size is consistent.

>This is not expected. Components should be unmounted and remounted on scroll.
Furthermore, tiles can be wrapped with memo to ensure we don't need to re-render.

As discussed, Added `React.memo` to the renderTile component to prevent unnecessary re-renders.

> Missing padding at the bottom.

fixed this. used padding for each cell instead of subtracting a constant from height and width to create space between tiles

Updated Screenshots
<img width="1496" alt="Screenshot 2020-07-08 at 12 05 32 AM" src="https://user-images.githubusercontent.com/9278015/86827081-31785180-c0af-11ea-8b92-018cda873cbb.png">
<img width="1499" alt="Screenshot 2020-07-08 at 12 05 56 AM" src="https://user-images.githubusercontent.com/9278015/86827092-35a46f00-c0af-11ea-9661-3437723b8af2.png">

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/Grid.tsx:
```diff
@@ -0,0 +1,44 @@
+  const columnCount = Math.max(1, itemCount && Math.ceil(itemCount / rowCount));
+  const cellRenderer = (data: GridCellProps) => children({ data, columnCount, items, rowCount });
+  return (
+    <GridComponent
```
Remove tab index from Grid:
```suggestion
    <GridComponent
        tabIndex={null}
```

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/Grid.scss:
```diff
@@ -0,0 +1,6 @@
+.ocs-grid {
+  .ReactVirtualized__Grid__innerScrollContainer {
+    width: 100% !important;
+    max-width: 100% !important;
```
Why is this needed? I removed this and it worked.

sahil143 replied:
This was added to give inner container full width, but it's not necessary anymore. Removed it.

christianvogt commented on frontend/packages/console-shared/src/components/virtualized-grid/VirtualizedGrid.tsx:
```diff
@@ -0,0 +1,94 @@
+                    {isItemsGrouped ? (
+                      <GroupByFilterGrid {...props} width={width} items={items as GroupedItems}>
+                        {(gridProps: GridChildrenProps) => (
+                          <Cell
+                            {...gridProps}
+                            renderHeader={renderHeader}
+                            renderCell={renderCell}
```
I wonder if this is where the lack of a `key` warning is coming from?

sahil143 replied:
yes this was the cause for key warning in console.

sahil143 commented:
> I see this console:
> ![image](https://user-images.githubusercontent.com/14068621/86962806-780e9000-c131-11ea-95c9-25a715aeba7d.png)
> ![image](https://user-images.githubusercontent.com/14068621/86963941-531b1c80-c133-11ea-8b96-b55705f71e4f.png)
> 
> When using the type filters, even when there's enough space for more columns, sometimes it shows less:
> ![image](https://user-images.githubusercontent.com/14068621/86962987-c328a300-c131-11ea-998d-ebcdaae96fbb.png)

@christianvogt fixed both the issue.

updated screenshot
<img width="1790" alt="Screenshot 2020-07-09 at 6 18 55 PM" src="https://user-images.githubusercontent.com/9278015/87042022-b3cc5700-c210-11ea-9944-8594938b8cd4.png">

christianvogt commented:
/lgtm
/approve

rohitkrai03 commented:
/retest

rohitkrai03 commented:
@sahil143 It would also be good if you could squash your commits into better explanatory commit messages around the feature.
